File: uretim_planlama/public/js/import_pre_check.js
// Import Öncesi Stok Kontrol Sistemi

frappe.ui.form.on('Data Import', {
    refresh: function(frm) {
        // Import butonu yanına "Stok Kontrolü" butonu ekle
        if (frm.doc.reference_doctype === 'Profile Stock Ledger') {
            frm.add_custom_button(__('Sayım Öncesi Kontrol'), function() {
                check_existing_stock_before_import(frm);
            }, __('Araçlar'));
        }
    }
});

function check_existing_stock_before_import(frm) {
    if (!frm.doc.import_file) {
        frappe.msgprint(__('Önce dosya yükleyiniz.'));
        return;
    }
    
    // Loading göster
    let d = new frappe.ui.Dialog({
        title: __('Sayım Öncesi Stok Kontrolü'),
        fields: [
            {
                fieldtype: 'HTML',
                fieldname: 'loading',
                options: `
                    <div class="text-center" style="padding: 20px;">
                        <div class="spinner-border" role="status">
                            <span class="sr-only">Yükleniyor...</span>
                        </div>
                        <p class="mt-2">Stok kontrolü yapılıyor...</p>
                    </div>
                `
            }
        ],
        primary_action_label: __('Kapat'),
        primary_action: function() {
            d.hide();
        }
    });
    d.show();
    
    // CSV dosyasını oku ve analiz et
    frappe.call({
        method: 'frappe.client.get',
        args: {
            doctype: 'File',
            name: frm.doc.import_file
        },
        callback: function(response) {
            if (response.message) {
                analyze_csv_content(response.message.file_url, d);
            }
        }
    });
}

function analyze_csv_content(file_url, dialog) {
    // CSV içeriğini analiz et (basit örnek)
    frappe.call({
        method: 'uretim_planlama.api.import_pre_check.check_existing_stock_before_import',
        args: {
            import_data: [] // CSV içeriği buraya parse edilecek
        },
        callback: function(response) {
            if (response.message && response.message.status === 'success') {
                show_import_analysis(response.message, dialog);
            } else {
                show_error_message(response.message, dialog);
            }
        }
    });
}

function show_import_analysis(data, dialog) {
    let conflicts_html = '';
    let new_items_html = '';
    
    // Çakışmalar
    if (data.conflicts && data.conflicts.length > 0) {
        conflicts_html = `
            <div class="alert alert-warning">
                <h5>⚠️ Stok Çakışmaları Bulundu (${data.conflicts.length} adet)</h5>
                <div class="table-responsive">
                    <table class="table table-sm">
                        <thead>
                            <tr>
                                <th>Ürün Kodu</th>
                                <th>Boy</th>
                                <th>Mevcut Stok</th>
                                <th>Import Miktarı</th>
                                <th>Toplam (Sonrası)</th>
                            </tr>
                        </thead>
                        <tbody>
        `;
        
        data.conflicts.forEach(function(conflict) {
            conflicts_html += `
                <tr>
                    <td>${conflict.item_code}</td>
                    <td>${conflict.length}</td>
                    <td>${conflict.existing_stock}</td>
                    <td>${conflict.import_qty}</td>
                    <td><strong>${conflict.total_after_import}</strong></td>
                </tr>
            `;
        });
        
        conflicts_html += `
                        </tbody>
                    </table>
                </div>
            </div>
        `;
    }
    
    // Yeni ürünler
    if (data.new_items && data.new_items.length > 0) {
        new_items_html = `
            <div class="alert alert-success">
                <h5>✅ Yeni Ürünler (${data.new_items.length} adet)</h5>
                <div class="table-responsive">
                    <table class="table table-sm">
                        <thead>
                            <tr>
                                <th>Ürün Kodu</th>
                                <th>Boy</th>
                                <th>Import Miktarı</th>
                            </tr>
                        </thead>
                        <tbody>
        `;
        
        data.new_items.forEach(function(item) {
            new_items_html += `
                <tr>
                    <td>${item.item_code}</td>
                    <td>${item.length}</td>
                    <td>${item.import_qty}</td>
                </tr>
            `;
        });
        
        new_items_html += `
                        </tbody>
                    </table>
                </div>
            </div>
        `;
    }
    
    // Özet
    let summary_html = `
        <div class="alert alert-info">
            <h5>📊 Import Özeti</h5>
            <ul class="mb-0">
                <li>Toplam Kayıt: ${data.summary.total_rows}</li>
                <li>Çakışmalı: ${data.summary.conflict_count}</li>
                <li>Yeni Ürün: ${data.summary.new_count}</li>
                <li>Çakışmalı Miktar: ${data.summary.total_conflict_qty}</li>
                <li>Yeni Miktar: ${data.summary.total_new_qty}</li>
            </ul>
        </div>
    `;
    
    // Dialog içeriğini güncelle
    dialog.fields_dict.loading.$wrapper.html(`
        <div style="padding: 20px;">
            ${summary_html}
            ${conflicts_html}
            ${new_items_html}
            
            ${data.summary.has_conflicts ? `
                <div class="alert alert-danger">
                    <h5>🚨 Dikkat!</h5>
                    <p>Bazı ürünler için mevcut stok bulundu. Import işlemi bu stoklara ekleme yapacak.</p>
                    <p><strong>Devam etmek istediğinizden emin misiniz?</strong></p>
                </div>
            ` : `
                <div class="alert alert-success">
                    <h5>✅ Güvenli Import</h5>
                    <p>Hiçbir stok çakışması bulunamadı. Import işlemi güvenle devam edebilir.</p>
                </div>
            `}
        </div>
    `);
}

function show_error_message(error, dialog) {
    dialog.fields_dict.loading.$wrapper.html(`
        <div class="alert alert-danger" style="padding: 20px;">
            <h5>❌ Hata</h5>
            <p>Stok kontrolü sırasında hata oluştu:</p>
            <p><code>${error.message || 'Bilinmeyen hata'}</code></p>
        </div>
    `);
}

// Import sonrası özet rapor
frappe.ui.form.on('Data Import', {
    after_import: function(frm) {
        if (frm.doc.reference_doctype === 'Profile Stock Ledger') {
            show_import_summary_report();
        }
    }
});

function show_import_summary_report() {
    frappe.call({
        method: 'uretim_planlama.api.import_pre_check.get_import_summary_report',
        callback: function(response) {
            if (response.message && response.message.status === 'success') {
                let d = new frappe.ui.Dialog({
                    title: __('Import Sonrası Özet'),
                    fields: [
                        {
                            fieldtype: 'HTML',
                            fieldname: 'summary',
                            options: generate_summary_html(response.message.data)
                        }
                    ],
                    primary_action_label: __('Kapat'),
                    primary_action: function() {
                        d.hide();
                    }
                });
                d.show();
            }
        }
    });
}

function generate_summary_html(data) {
    let html = `
        <div style="padding: 20px;">
            <div class="alert alert-info">
                <h5>📊 Son 30 Gün İstatistikleri</h5>
                <ul class="mb-0">
                    <li>Toplam Profile Entry: ${data.statistics.total_recent_entries}</li>
                    <li>Toplam Miktar: ${data.statistics.total_recent_qty}</li>
                    <li>Dönem: ${data.statistics.period}</li>
                </ul>
            </div>
        </div>
    `;
    return html;
}

File: uretim_planlama/uretim_planlama/api/import_pre_check.py
import frappe
from frappe import _
from frappe.utils import flt
from uretim_planlama.uretim_planlama.utils import get_profile_stock

@frappe.whitelist()
def check_existing_stock_before_import(import_data):
	"""
	Sayım güncellemesi öncesi mevcut stokları kontrol et
	
	Args:
		import_data: [{"item_code": "123", "length": "5.0", "qty": 10}, ...]
	
	Returns:
		{
			"stock_changes": [...],  # Stok değişiklikleri
			"new_items": [...],      # Yeni ürünler
			"no_change": [...],      # Değişiklik olmayan ürünler
			"summary": {...}         # Özet bilgiler
		}
	"""
	try:
		if isinstance(import_data, str):
			import_data = frappe.parse_json(import_data)
		
		stock_changes = []
		new_items = []
		no_change = []
		total_increase_qty = 0
		total_decrease_qty = 0
		total_new_qty = 0
		
		for row in import_data:
			item_code = row.get("item_code")
			length = str(row.get("length", ""))
			qty = flt(row.get("qty", 0))
			
			if not item_code or not length or qty < 0:
				continue
			
			# Mevcut stok kontrolü
			existing_stock = get_profile_stock(item_code, length)
			stock_difference = qty - existing_stock
			
			if existing_stock > 0:
				if stock_difference > 0:
					# Stok artışı
					stock_changes.append({
						"item_code": item_code,
						"length": length,
						"existing_stock": existing_stock,
						"new_stock": qty,
						"difference": stock_difference,
						"change_type": "increase",
						"info": f"📈 Stok artışı: {existing_stock} → {qty} (+{stock_difference})"
					})
					total_increase_qty += stock_difference
				elif stock_difference < 0:
					# Stok azalışı
					stock_changes.append({
						"item_code": item_code,
						"length": length,
						"existing_stock": existing_stock,
						"new_stock": qty,
						"difference": abs(stock_difference),
						"change_type": "decrease",
						"info": f"📉 Stok azalışı: {existing_stock} → {qty} (-{abs(stock_difference)})"
					})
					total_decrease_qty += abs(stock_difference)
				else:
					# Değişiklik yok
					no_change.append({
						"item_code": item_code,
						"length": length,
						"stock": existing_stock,
						"info": "✅ Değişiklik yok"
					})
			else:
				# Yeni ürün
				new_items.append({
					"item_code": item_code,
					"length": length,
					"new_stock": qty,
					"info": "🆕 Yeni ürün"
				})
				total_new_qty += qty
		
		summary = {
			"total_rows": len(import_data),
			"stock_changes_count": len(stock_changes),
			"new_items_count": len(new_items),
			"no_change_count": len(no_change),
			"total_increase_qty": total_increase_qty,
			"total_decrease_qty": total_decrease_qty,
			"total_new_qty": total_new_qty,
			"has_changes": len(stock_changes) > 0
		}
		
		return {
			"stock_changes": stock_changes,
			"new_items": new_items,
			"no_change": no_change,
			"summary": summary,
			"status": "success"
		}
		
	except Exception as e:
		frappe.log_error(f"Import pre-check error: {str(e)}", "Import Pre-check Error")
		return {
			"status": "error",
			"message": str(e)
		}

@frappe.whitelist()
def get_import_summary_report():
	"""Import öncesi özet rapor oluştur"""
	try:
		# Son 30 gün içindeki Profile Entry'leri al
		recent_entries = frappe.get_all(
			"Profile Entry",
			filters={
				"creation": [">=", frappe.utils.add_days(frappe.utils.today(), -30)],
				"docstatus": 1
			},
			fields=["name", "creation", "total_received_qty"],
			order_by="creation desc",
			limit=50
		)
		
		# Son 30 gün içindeki Profile Stock Ledger değişikliklerini al
		recent_ledger_changes = frappe.get_all(
			"Profile Stock Ledger",
			filters={
				"creation": [">=", frappe.utils.add_days(frappe.utils.today(), -30)]
			},
			fields=["item_code", "length", "qty", "creation"],
			order_by="creation desc",
			limit=100
		)
		
		# İstatistikler
		total_recent_entries = len(recent_entries)
		total_recent_qty = sum(flt(entry.total_received_qty) for entry in recent_entries)
		
		return {
			"status": "success",
			"data": {
				"recent_entries": recent_entries[:10],  # Son 10 giriş
				"recent_ledger_changes": recent_ledger_changes[:10],  # Son 10 değişiklik
				"statistics": {
					"total_recent_entries": total_recent_entries,
					"total_recent_qty": total_recent_qty,
					"period": "Son 30 gün"
				}
			}
		}
		
	except Exception as e:
		frappe.log_error(f"Import summary report error: {str(e)}", "Import Summary Error")
		return {
			"status": "error",
			"message": str(e)
		}
